### ISSUE TITLE
Changing the values of a FieldArray using change action creator after a field has received focus throws exception

### ISSUE THREAD

**Issue by sargant** [erikras/redux-form#2058]
When you attempt to replace an entire array of data (using the `change` action creator) used by a `FieldArray` component after an element from the existing dataset has been focused, an exception is thrown with the message **`Cannot delete non-numerical index from an array`**

I've made some changes to the FieldArray example project that can be used to replicate the bug - the changes are in commit https://github.com/sargant/redux-form/commit/497f01e1878dee33932b26f4fb8d0bac60041fc4 on my fork.

In order to replicate the bug:

* Click "Value set A" to populate the form with a predefined set of data
* Focus and blur any field in the `FieldArray` part of the form, e.g. a first name
* Click "Value set B" to populate the form with a different predefined set of data

The error is thrown instead of the data set changing. If you don't focus a field before changing the data set, it works fine.

**Comment by jmaltis** [comment]
Same problem here

**Comment by imthehell** [comment]
Same problem also to me...

**Comment by sargant** [comment]
My current workaround is to change each field individually in a loop, and then if list B is shorter than list A, use `arrayRemove` to delete the items that won't have been replaced. Hacky, but it works.

```javascript
// get the lengths of the two sets of items
let oldItemsLength = this.props.formValues.members.length;
let newItemsLength = newItems.length;

// change each item individually
newItems.map((item, i) => this.props.RxF.change(formName, `members[${i}]`, item));

// manually remove any items if the new list is shorter than the old list
for (let i = newItemsLength; i < oldItemsLength; i++) {
  this.props.RxF.arrayRemove(formName, `members[${i}]`, i);
}
```

**Comment by jmaltis** [comment]
Thank you, I made it work using the almost the same solution.

Instead of remove one by one the missing elements of the array, I removed them all in once thanks to : 
`this.props.RxF.removeAll(`members`);`. That way, the form is re rendered just once when removing. Then, I pushed the new values with `newItems.map((newItem) => rxf.array.push(`members `, newItem));` 

PS : by the way, you have a typo in your last line -> 
`this.props.RxF.arrayRemove(formName, `members[${i}]`, i);` instead of `this.props.RxF.arrayRemove(formName, `members`, i);`
        // Add all new grades
        grades.map((grade) => dispatch(normalize(rxf.array.push(`assessmentGrades`, grade))));

**Comment by sargant** [comment]
Thanks @jmaltis - trying to anonymise the code and messed up! Change it.

**Comment by davincho** [comment]
I can confirm this bug:

Throws an `Cannot delete non-numerical index from an array` error, when one of the fields has been focused before:
```javascript
change('prices', defaultValues.prices);
```

Works:
```javascript
array.removeAll('prices');
defaultValues.prices.forEach(price => array.push('prices', price));
```

**Comment by swp44744** [comment]
+1 Facing same issue.. @erikras Any plans to fix this bug in upcoming releases? change action is really helpful in big applications. it would be great if we can address this issue.

**Comment by imthehell** [comment]
Any news about the fixing of this bug?

**Comment by mqklin** [comment]
This works for me:
```
myChange: (array, idx, value) => {
  const changedElement = { ...array[idx], field_name: value };
  dispatch(arrayRemove(FORM_NAME, ARRAY_NAME, idx));
  dispatch(arrayInsert(FORM_NAME, ARRAY_NAME, idx, changedElement));
},
```

**Comment by sinslav** [comment]
+1

**Comment by matmiz** [comment]
I'm experiencing the same issue, but the solutions above are not helping me. 
The dispatching of remove and insert are causing side effects in my code, and I can't seem to solve them.
Is there any chance this bug will be fixed soon?

Thanks

**Comment by hoppy-kamper** [comment]
@matmiz post a Stack Overflow question and I'd be happy to take a look for you. We should be able to find a workaround.

**Comment by hoanglong10026** [comment]
I using react-select + redux form and meet same issue.
This work for me:

```
addTag(value){
        const values = [...this.props.form_data.tags, {value: value, label: value}];
        this.props.dispatch(change('GroupForm', 'tags', values));
}
```

**Comment by gustavohenke** [comment]
Because this issue saw almost no activity for a few months, I'm closing it.

**Comment by webmobiles** [comment]
any new about this function?

**Comment by shkdmhhb** [comment]
Commented in #3626, there seems to be another scenario where the above error occurs when using clearFields on an array field.

**Comment by ArtanisAce** [comment]
It's solved in 7.3.0, I can confirm also. Pity to see nobody from the dev team showed up :/
